Add VQ-BeT

jayLEE0301 opened the pull request:
## What this does
Add [VQ-BeT](https://sjlee.cc/vq-bet) for PushT env.

## How it was tested
Explain/show how you tested your changes.

Examples:
- Added `configuration_vqbet.py` and `modeling_vqbet.py` in `vqbet` folder.

## How to checkout & try? (for the reviewer)

Examples:
```bash
python lerobot/scripts/train.py policy=vqbet env=pusht dataset_repo_id=lerobot/pusht
```

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/huggingface/lerobot/166)
<!-- Reviewable:end -->

alexander-soare reviewed (changes requested):
@jayLEE0301 thanks so much for being the first to PR a model to LeRobot! The paper for VQ-BeT was a really nice read.

So, for the review. I've left a bunch of comments (many of them nits, but some blockers), and actually decided to stop reviewing partway through. That's because I noticed there are some high-level points I can share here. So instead, please consider these high level comments as my primary review, and my inline comments as examples to support.

So, our goal is to make this code **highly accessible** to the community, meaning it's **easy to read and understand**, and is **easily hackable**. A side effect of aiming for these goals is usually that the code is maintainable.

With that overarching goal in mind here are 3 high level points:

1. Consider the `VQBeTPolicy` class as the only "public" object in the modeling file. Everything else is there for the sole purpose of `VQBeTPolicy`. This means:
     - **Go minimalist**. We should drop any kwargs, conditional branching, or other logic that is unused. The other functions and logic should only be as dynamic as needed to serve `VQBeTPolicy`. Rule of thumb: if it can't be be activated via the configuration parameters, it can go
    - **Use the `config` instead of many kwargs**. Most of the other modules can take a `config` argument and make a `self.config` (avoids relisting parameters twice, and makes it that there's one source of truth for what the params mean - no need to repeat documentation or type-hinting).
   
 2. **Consolidate code**: We want to avoid too much nesting or duplication of code. Consider for example my inline comment about the MLPs. I think it's reasonable to use one class for MLPs (and it can be simpler and shorter than the 3 existing classes now). This is just an example though, there may be more opportunities for consolidation.
 
 3. **Documentation and naming**: We want to make sure that everything is well understood by a first-time reader. Wear the hat of someone who has read through your paper once, and enters the code via the `VQBeTPolicy` class. They should be able to traverse the submodule hierarchy, understanding what everything is as they go. And they should be able to make sense of what's happening in the forward function.
     - Above all, please make sure the `VQBeTConfig` documentation is solid.
     - Please add docstrings to classes and methods when it wouldn't be obvious what they are in relation to the main policy and paper.
     - Please separate long methods into logical blocks with comments so that one doesn't get lost along they way. (btw: this doesn't mean separating them into smaller functions)
     - Please make sure it's easy to follow what's happening with tensor dimensions.  `einops` is also helpful for that.
     -  Favor full words over abbreviations: `embd` -> `embed` and try to match the terminology/naming in your paper.
  
  When in doubt, please take inspiration from LeRobot's ACT and TD-MPC (Diffusion Policy is good too but may need a little more work).

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+        self.ln_1 = nn.LayerNorm(config.n_embd)
+        self.attn = CausalSelfAttention(config)
+        self.ln_2 = nn.LayerNorm(config.n_embd)
+        self.mlp = GPT_MLP(config)
```
Good idea. I made the changes you recommended.

jayLEE0301 commented on lerobot/configs/policy/vqbet.yaml:
```diff
@@ -0,0 +1,99 @@
+# @package _global_
+
+# Defaults for training for the PushT dataset as per https://github.com/real-stanford/diffusion_policy.
+# Note: We do not track EMA model weights as we discovered it does not improve the results. See
```
Deleted.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+        else:
+            return loss.sum()
+
+class MLP(torch.nn.Sequential):
```
Removed this MLP class , and Integrated with other MLPs.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+
+#     def __getitem__(self, idx):
+#         return self.actions[idx]
+class EncoderMLP(nn.Module):
```
removed.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+        self.n_action_steps = n_action_steps # action chunk size
+
+
+        if self.sequentially_select:
```
Actually, we do not use self.sequentially_select in all envs, so removed this option.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+        hidden_size=1024,
+        # loss_kwargs
+        offset_loss_weight=100.0,
+        secondary_code_multiplier=0.5,
```
Changed the name `secondary_code_multiplier` -> `secondary_code_loss_weight`

jayLEE0301 commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,133 @@
+    Args:
+        n_obs_steps: Number of environment steps worth of observations to pass to the policy (takes the
+            current step and additional steps going back).
+        horizon: Diffusion model action prediction size as detailed in `DiffusionPolicy.select_action`.
```
removed unused parameter explanations.

jayLEE0301 commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,133 @@
+        use_group_norm: Whether to replace batch normalization with group normalization in the backbone.
+            The group sizes are set to be about 16 (to be precise, feature_dim // 16).
+        spatial_softmax_num_keypoints: Number of keypoints for SpatialSoftmax.
+        down_dims: Feature dimension for each stage of temporal downsampling in the diffusion modeling Unet.
```
^

jayLEE0301 commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,133 @@
+        down_dims: Feature dimension for each stage of temporal downsampling in the diffusion modeling Unet.
+            You may provide a variable number of dimensions, therefore also controlling the degree of
+            downsampling.
+        kernel_size: The convolutional kernel size of the diffusion modeling Unet.
```
^

jayLEE0301 commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,133 @@
+    vqvae_embedding_dim: int = 256
+    # VQ-BeT
+    block_size: int = 50
+    output_dim: int = 256
```
added `gpt_` in front of the names, to make it clear that they are used in `class GPT`

jayLEE0301 commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,133 @@
+    output_dim: int = 256
+    n_layer: int = 6
+    n_head: int = 6
+    n_embd: int = 120
```
I thought that `dim_model` could be confusing, since there's multiple models (GPT, VQ-VAE, header ...). Would it be ok to set it as `gpt_hidden_dim`?

alexander-soare replied:
Sounds good :)

jayLEE0301 commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,133 @@
+    n_head: int = 6
+    n_embd: int = 120
+    dropout: float = 0.1
+    mlp_hidden_dim: int = 1024
```
I think that `mlp_hidden_dim` might be better, in terms of consistency with `gpt_hidden_dim` and `vqvae_enc_hidden_dim`

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+        queue is empty.
+        """
+
+        # seungjae TODO: implement averaging action over horizons
```
I tested it, but seems similar or worse than just not averaging over the multiple action. (and even require more computation power, since it should predict action chunk every timestep) So determined to drop it.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+        self._obs_queues = populate_queues(self._obs_queues, batch)
+
+        if not self.check_discretized():
+            raise NotImplementedError(
```
This error is called if evaluation is attempted before the end of discretisation. Added a related comment.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+        # for act seq pred, we should provide averaged act over horizon.
+
+        if len(self._action_queue) == 0:
+            # `self.model.forward` returns a (batch_size, n_action_steps, action_dim) tensor, but the queue
```
Changed the location of comment and added detailed explanations

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+        self.rgb_encoder = DiffusionRgbEncoder(config)
+
+        global_cond_dim = (config.output_shapes["action"][0] + self.rgb_encoder.feature_dim)
+        self._policy = GPT(
```
changed the modules (self.policy, self.action_head) not include `_`

jayLEE0301 replied:
removed `_` in all the pytorch model classes

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+    def discretize(self, discretize_step, actions):
+        return self._action_head.discretize(discretize_step, actions)
+
+    # ========= inference  ============
```
dropped the comment

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+        # Concatenate state and image features then flatten to (B, global_cond_dim).
+        global_cond = torch.cat([batch["observation.state"], img_features], dim=-1)
+
+        obs = global_cond
```
good idea! changed the name as suggested.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+            action = batch["action"]
+        else:
+            action = None
+        if 'goal' in batch.keys():
```
It was some remnant for "conditional VQ-BeT", but since we are serving only unconditional one now, we dropped it. But will implement it again after implementing some goal-cond envs.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+
+
+class VQBeTHead(nn.Module):
+    def __init__(
```
removed kwargs, and makes it use only self.config.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2902 @@
+            return pred_action, loss[0] if isinstance(loss, tuple) else loss
+
+
+class VQBeTHead(nn.Module):
```
add some explanations in this part

jayLEE0301 commented:
Thank you for the review!

Following these high-level points,

1. we removed all kwargs as possible, and only remained configs.
2. consolidated all the similar functions
3. added comments, and changed names

for all the parts of this PR.





<!-- Sent from Reviewable.io -->

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        if rollout:
+            return pred_action["predicted_action"][:, n_obs_steps-1, :].reshape(batch_size, self.config.n_action_pred_chunk, -1)
+        # else, it calculate overall loss (bin prediction loss, and offset loss)
+        else:
+            action = batch["action"]
+            n, total_w, act_dim = action.shape
+            act_w = self.config.n_action_pred_chunk
+            num_token = total_w + 1 - act_w
+            output_shape = (n, num_token, act_w, act_dim)
+            output = torch.empty(output_shape).to(action.device)
+            for i in range(num_token):
+                output[:, i, :, :] = action[:, i : i + act_w, :]
+            action = output
+
+            loss = self.action_head.loss_fn(
+                pred_action,
+                action,
+                reduction="mean",
+            )
```
Do you mind if we simplify here?

First we need this in the `__init__`:

```python
num_tokens = self.config.n_action_pred_token + self.config.n_action_pred_chunk - 1
self.register_buffer(
                "select_target_actions_indices",
                torch.row_stack(
                    [torch.arange(i, i + self.config.n_action_pred_chunk) for i in range(num_tokens)]
                ),
            )
```

This saves us from recreating the tensor each forward pass. We actually have a performance related task in to do this with ACT https://github.com/orgs/huggingface/projects/46?pane=issue&itemId=64971003.

Then here we can cut down a lot (might need some verification on your part):

```suggestion
        else:
            output = batch["action"][:, self.select_target_actions_indices]
            loss = self.action_head.loss_fn(pred_action, output, reduction="mean")
            return pred_action, loss
```

alexander-soare replied:
Btw it's also okay to create the indices on the spot if you prefer that to using a buffer (it might be more readable). But they should be created with `device=batch['action'].device` in order to create them directly on GPU rather than creating them on CPU (because then we need a device transfer).

alexander-soare replied:
FYI here's a thread where I'm trying to understand the performance implications: https://discord.com/channels/1216765309076115607/1237737604611571722/1246144412329709578. But I would still say my suggestion above mostly stands because I think it makes the code easier to read.

jayLEE0301 replied:
Thank you for your suggestion! I changed following your code.

alexander-soare reviewed (changes requested):
Checkpoint. I will continue next week.

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        offset_loss = torch.nn.L1Loss()(action_seq, predicted_action)
+
+        # calculate primary code prediction loss
+        cbet_loss1 = self._criterion(  # F.cross_entropy
```
The comment says `F.cross_entropy` but `self._criterion` is `FocalLoss`. To avoid doubt can we please name it `self._focal_loss_fn` and remove the comments?

jayLEE0301 replied:
fixed:)

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+    return loss, n_different_codes, n_different_combinations
+
+
+def exists(val):
```
This appears to be defined twice. Also, is it okay if we avoid using these simple functions that wrap short 1-liners? It adds a layer of indirection which would make it harder to read the code (reader has to go and check what `exists` does).

jayLEE0301 replied:
deleted the parts that were defined twice and some short functions. (`def default`, `def exists`, `def round_up_multiple`, and `def l2norm`) But I left some of the functions that is used in the very low-level part of RVQ since removing all these short function in low-level of RVQ doesn't necessarily lead to more readable code.

alexander-soare replied:
Excellent. Thank you :)

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+
+        loss, n_different_codes, n_different_combinations = pretrain_vqvae(self.vqvae_model, discretize_step, actions)
+        # if we updated RVQ more than `discretize_step` steps,
+        if self.vqvae_model.discretized:
```
Any chance we can move this logic to the top of the stack in `VQBeTPolicy`? There is already a similar check there, and I think it would be clearer to have all this logic in one place.

By the way, the call to `eval` will be soon undermined by the training loop which will call `train` on the top level module. Is that okay? If so, should we remove it here?

jayLEE0301 replied:
This part is just for setting vqvae_model as `eval()`.

I moved this part inside `def pretrain_vqvae(vqvae_model, discretize_step, actions):`
Would it make sense?

alexander-soare replied:
Thanks for that! I'll double check when I get to it :)

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+            action_seq = action_seq.unsqueeze(0)
+
+        # offset loss is L1 distance between the predicted action and ground truth action
+        offset_loss = torch.nn.L1Loss()(action_seq, predicted_action)
```
nit: Use functional rather than instantiating a module:

```suggestion
        offset_loss = F.l1_loss(action_seq, predicted_action)
```

jayLEE0301 replied:
fixed

alexander-soare commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,132 @@
+    spatial_softmax_num_keypoints: int = 32
+    # VQ-VAE
+    discretize_step: int = 3000
+    vqvae_groups: int = 2
```
In some places of the code, this is statically handled, meaning changing this number will break things. Can we please either remove it as a parameter or make sure the code can handle it dynamically?

One example is the cbet_loss.

jayLEE0301 replied:
made the code can handle `vqvae_groups` dynamically

jayLEE0301 replied:
Oh I got it. I thought that I made it can handle `vqvae_groups` dynamically, but if we turn on `sequentially_select=True`, it still can only handle `vqvae_groups=2`.

There's some way to make it can handle `vqvae_groups` dynamically, but I would like to make it fixed as `vqvae_groups=2`, since it is has never changed across all the envs in the original paper.
(FYI, `sequentially_select=True` is not used in PushT, but in all the real-world experiments with hello robot in the our manuscript, we used `sequentially_select=True`, so we are maintaining that option.)

In summary, we decided to remove `vqvae_groups ` option and just fix it as 2.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+
+from lerobot.common.policies.vqbet.configuration_vqbet import VQBeTConfig
+from huggingface_hub import PyTorchModelHubMixin
+from robomimic.models.base_nets import SpatialSoftmax
```
changed to use SpatialSoftmax from modeling_diffusion.py

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+
+
+        # queues are populated during rollout of the policy, they contain the n latest observations and actions
+        self._queues = None
```
added self.reset() at the bottom of `__init__`

alexander-soare replied:
Any chance we can also drop `self._queues = None` please? It doesn't hurt from a logic perspective, but it does potentially confuse someone who will wonder why there's a redundant line of code here.

jayLEE0301 replied:
deleted this line:)

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+
+        self.vqbet = VQBeTModel(config)
+
+    def check_discretized(self):
```
removed and fixed to use directly

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        if not self.check_discretized():
+            self.vqbet.action_head.vqvae_model.discretized = True
+            # VQ-BeT can predict action only after finishing action discretization.
+            # We added a logit to force self.vqbet.action_head.vqvae_model.discretized to be True if not self.check_discretized() to account for the case of predicting with a pretrained model, but this shouldn't happen if you're learning from scratch, so set eval_freq greater than discretize_step.
```
removed this part, and added self.register_buffer('discretized', torch.tensor(False)) in class VqVae

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+
+            # the dimension of returned action is (batch_size, n_action_pred_chunk, action_dim)
+            actions = self.unnormalize_outputs({"action": actions})["action"]
+            # since the data in the action queue's dimension is (n_action_pred_chunk, batch_size, action_dim, we transpose the action and fill the queue
```
fixed

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        # if Residual VQ is already trained, VQ-BeT trains its GPT and bin ped header / offset header parts.
+        _, loss = self.vqbet(batch, rollout=False)
+
+        return loss
```
actually it was returning loss_dict (which is the output of `def loss_fn`) to clarify, I changed the output name as loss_dict

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        - These `features` pass through the action head, which passes through the code prediction, offset prediction head, 
+        and finally generates a prediction for the action chunks.
+
+        -------------------------------** legend **-------------------------------
```
😊

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        # VQ-BeT discretizes action using VQ-VAE before training BeT (please refer to section 3.2 in the VQ-BeT paper https://arxiv.org/pdf/2403.03181)
+        if not self.check_discretized():
+            loss, n_different_codes, n_different_combinations = self.vqbet.discretize(self.config.discretize_step, batch['action'])
+            return {"loss": loss, "n_different_codes": n_different_codes, "n_different_combinations": n_different_combinations}
```
added explanations:)

alexander-soare replied:
Thanks! I _think_ I understand this. Can you let me know if my understanding is correct?

```
n_different_codes: how many of the total possible VQ codes are being used (how many of them have at least one encoder embedding as a nearest neighbor). This can be at most `vqvae_n_embed`.
n_different_combinations: how many different code combinations are being used out of all possible combinations. This can be at most `vqvae_n_embed ^ vqvae_groups` (hint consider the RVQ as a decision tree).

But shouldn't `n_different_codes` max out at `vqvae_n_embed * vqvae_groups`? That's how many codes there are in total. Or are you only referring to the codes of the first RVQ layer?

Btw: I think this is a great metric to track!

jayLEE0301 replied:
Exactly. The reason why the max value of n_different_codes is `vqvae_n_embed` is we don't differentiated the code of different RVQ layers. But after your comment, I agree that we should differentiate them to calculate `n_different_codes`. Thus, I changed the `n_different_codes` calculation part. 

from
```
n_different_codes = len(torch.unique(metric[2]))
```
to
```
n_different_codes = sum([len(torch.unique(metric[2][:, i])) for i in range(vqvae_model.vqvae_num_layers)])
```

Also, thanks to your more detailed explanations, I updated the comments.

```
# n_different_codes: how many of the total possible VQ codes are being used in single batch (how many of them have at least one encoder embedding as a nearest neighbor). This can be at most `vqvae_n_embed * number of layers of RVQ (=2)`.
# n_different_combinations: how many different code combinations are being used out of all possible combinations in single batch. This can be at most `vqvae_n_embed ^ number of layers of RVQ (=2)` (hint consider the RVQ as a decision tree).
```

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        if not self.check_discretized():
+            loss, n_different_codes, n_different_combinations = self.vqbet.discretize(self.config.discretize_step, batch['action'])
+            return {"loss": loss, "n_different_codes": n_different_codes, "n_different_combinations": n_different_combinations}
+        # if Residual VQ is already trained, VQ-BeT trains its GPT and bin ped header / offset header parts.
```
changed errors: "header" -> "head", "bin pred" -> "bin prediction head"

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        --------------------------------------------------------------------------
+
+        
+        Phase 1. Discretize action using Residual VQ (for config.discretize_step steps)
```
fixed Phase -> Training Phase

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+                    └── + ──┘         └── + ──┘                     └── + ──┘       └── + ──┘
+                        ▼                 ▼                             ▼               ▼
+                   action chunk      action chunk                  action chunk     action chunk
+                    a_{t_n+1} ~       a_{t_n+2} ~                   a_{t} ~     ...  a_{t+p-1} ~ 
```
Thank you! I fixed it.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+                self._action_token.repeat(batch_size, n_obs_steps, 1, 1)
+            ], dim=-2).view(batch_size, -1, self.config.gpt_input_dim)
+        if img_features.shape[1] != n_obs_steps:
+            raise NotImplementedError
```
changed raise error to `assert`

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+                config.output_shapes["action"][0], 
+                hidden_channels=[self.config.gpt_input_dim]
+            )
+        self.obs_projector = MLP(
```
Thank you for your suggestion! changed `obs_projector` to `rgb_feature_projector`

jayLEE0301 commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,132 @@
+        gpt_n_layer: Number of layers of GPT
+        gpt_n_head: Number of headers of GPT
+        gpt_hidden_dim: Size of hidden dimensions of GPT
+        gpt_num_obs_mode: Number of different observation modes. (e.g., PushT env: {state, image observation}, thus 2.)
```
Thank you for your suggestion, However, I think that if we leave this option, when we could add some only-state-input envs, or only-image-input envs later. How do you think about that?

alexander-soare replied:
I like the idea to do that in the future. We have plans to achieve this via the `input_shapes` dictionary (you can choose whether to put state and/or images), which is the "source of truth" for the inputs. `gpt_num_obs_mode` is redundant with `input_shapes`, right? What do you think of this approach?

jayLEE0301 replied:
Oh then we don't need `gpt_num_obs_mode`.
I removed it and now it only use `input_shapes`.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        # Separate batch and sequence dims.
+        img_features = einops.rearrange(img_features, "(b n) ... -> b n ...", b=batch_size)
+        
+        # image observation feature, state feature, and action query token are grouped together with the same timestpe to form a group, which is listed in order to be entered into GPT sequentially.
+        observation_feature = torch.cat([
+                torch.unsqueeze(self.obs_projector(img_features), dim=2), 
+                torch.unsqueeze(self.state_projector(batch["observation.state"]), dim=2), 
+                self._action_token.repeat(batch_size, n_obs_steps, 1, 1)
+            ], dim=-2).view(batch_size, -1, self.config.gpt_input_dim)
```
Thank you for your suggestion! Replaced the long line with your code:)

jayLEE0301 commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,132 @@
+        n_obs_steps: Number of environment steps worth of observations to pass to the policy (takes the
+            current step and additional steps going back).
+        n_action_pred_token: Number of future tokens that VQ-BeT predicts.
+        n_action_pred_chunk: Action chunk size of each aciton prediction token.
```
I think calling it `action_chunk_size` is better:) changed the name.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        self.rgb_encoder = VQBeTRgbEncoder(config)
+
+        # This action query token is used as a prompt for querying action chunks. Please refer to "A_Q" in the image above.
+        self._action_token = nn.Parameter(torch.randn(1, 1, self.config.gpt_input_dim))
```
This is because all of these tokens are query tokens that generate the "action chunk" (even though all the action chunk has different time of action). The positional embedding is added to this tokens.

Do you think it would make more sense to generate each token as a different random vector? (Actually, this was an issue we didn't have to consider in the original vq-bet, since we only used RGB input, and simply generated the action output from the RGB input token)

Let me test with different random vectors for action query token in different timestep.

jayLEE0301 replied:
I tested using different random vectors for each action query token, and it seems that it is worse than sharing all the same vectors. So I am planning to maintain current implementation.

alexander-soare replied:
Excellent. Thank you for that. To make sure your work is not forgotten or redone, what do you think of leaving a comment:

```
# Note: During the forward pass, this token is repeated as many times as needed. The authors also experimented with initializing the necessary number of tokens independently and observed inferior results.
```

jayLEE0301 replied:
Thank you:) I added the comment.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        features = self.policy(observation_feature)
+        historical_act_pred_index = np.arange(0, n_obs_steps) * (self.config.gpt_num_obs_mode+1) + self.config.gpt_num_obs_mode
+
+        # only extract the output tokens at the position of action query
```
This increases computation, but can help improve overall learning performance, and avoiding overfitting (not always)

You can think of it similar to predicting a longer sequence of actions in a diffusion policy compared to the actual sequence of actions to be performed.

alexander-soare replied:
Cool! Mind adding that in as a comment (if it's not already mentioned in your paper)?

jayLEE0301 replied:
I added
`Behavior Transformer (BeT), and VQ-BeT are both sequence-to-sequence prediction models, mapping sequential observation to sequential action (please refer to section 2.2 in BeT paper https://arxiv.org/pdf/2206.11251). Thus, we predict historical action sequence, in addition to current and future actions.`

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+            features[:, -len_additional_action_token:]
+        ], dim=1)
+        # pass through action head
+        pred_action = self.action_head(
+            features,
+        )
```
^

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        self._criterion = FocalLoss(gamma=2.0)
+
+    def discretize(self, discretize_step, actions):
+        if next(self.vqvae_model.encoder.parameters()).device != get_device_from_parameters(self):
```
I created it when `class VqVae` was not inheriting from `nn.module` (to manually match vqvqe_model's device with other part's device), but now that that's fixed, I've deleted it.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+            NT=NT,
+        )
+
+        indices = (
+            torch.arange(NT).unsqueeze(1).cuda(),
+            torch.arange(self.config.vqvae_groups).unsqueeze(0).cuda(),
+            sampled_centers,
+        )
```
Thanks! I checked and removed all .cuda() parts.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        )
+        # Use advanced indexing to sample the values
+        sampled_offsets = cbet_offsets[indices]
+        # Extract the only offsets corresponding to the sampled codes.
+        sampled_offsets = sampled_offsets.sum(dim=1)
```
changed the explanations! 

```
        # Use advanced indexing to sample the values (Extract the only offsets corresponding to the sampled codes.)
        sampled_offsets = cbet_offsets[indices]
        # Then, sum the offsets over the RVQ layers to get a net offset for the bin prediction
        sampled_offsets = sampled_offsets.sum(dim=1)
```

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        # Extract the only offsets corresponding to the sampled codes.
+        sampled_offsets = sampled_offsets.sum(dim=1)
+        # Get the centroids of each layer to pass it through RVQ decoder
+        centers = self.vqvae_model.draw_code_forward(sampled_centers).view(
+            NT, -1, self.config.vqvae_embedding_dim
+        )
+        return_decoder_input = einops.rearrange(
+            centers.clone().detach(), "NT 1 D -> NT D"
+        )
```
Thank you! changed as recommended.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        )  # action_bins: NT, G
+
+        # Now we can compute the loss.
+        if action_seq.ndim == 2:
```
This is not using. Deleted this part.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+            (action_bins[:, 1] == sampled_centers[:, 1]).int()
+        ) / (NT)
+
+        action_mse_error = F.mse_loss(
```
Deleted unnecessary things:)

jayLEE0301 commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,132 @@
+        use_group_norm: Whether to replace batch normalization with group normalization in the backbone.
+            The group sizes are set to be about 16 (to be precise, feature_dim // 16).
+        spatial_softmax_num_keypoints: Number of keypoints for SpatialSoftmax.
+        discretize_step: Number of optimization steps for training Residual VQ.
```
changed `discretize_step` -> `n_vqvae_training_steps`

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2382 @@
+        """
+        VQBeTHead takes output of GPT layers, and pass the feature through bin prediction head (`self.map_to_cbet_preds_bin`), and offset prediction head (`self.map_to_cbet_preds_offset`)
+
+        self.map_to_cbet_preds_bin: outputs probability of each code (for each layer).
```
Thanks!

alexander-soare reviewed (changes requested):
Just publishing my responses in a batch. Thanks for resolving these :D

Now moving on with the review.

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+            state = einops.rearrange(state, "N T A -> N (T A)")
+        return state
+
+    def get_code(self, state, required_recon=False):
```
This looks like it's never called with `required_recon=True` and there appears to be bout 10 lines of logic we might be able to do away with.

On another note, can you please take a look through the rest of the code and find any code that is not accessed? I think we should make a determination about whether it is relevant (now or in the _near_ future) or not, and if not, remove it. It makes the code easier to read and maintain.

jayLEE0301 replied:
Removed the unused parts. Thank you:)

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+        self.map_to_cbet_preds_bin: outputs probability of each code (for each layer).
+            The input dimension of `self.map_to_cbet_preds_bin` is same with the output of GPT, 
+            and the output dimension of `self.map_to_cbet_preds_bin` is `self.config.vqvae_groups * self.config.vqvae_n_embed`, where 
+                `self.config.vqvae_groups` is number of RVQ layers, and
```
I removed the duplicated parts:) Thank you

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+        }
+        return loss_dict
+
+class VQBeTOptimizer(nn.Module):
```
Thank you for the suggestions!
I removed all the redundant optimizers, and merged all the optimizers for phase 1 and phase 2 into one, leaving only one optimizer. I also deleted `def step`, `def zero_grad`.

We haven't done much analysis on how this affects the stability of training at this time, but (after running two seeds) we have found that it can produce similar performance to the uploaded model(https://huggingface.co/JayLee131/vqbet_pusht) based on the best checkpoint.

Perhaps a more diverse hyperparameter search may be needed in the future.

```
class VQBeTOptimizer(torch.optim.Adam):
    def __init__(self, policy, cfg):
        vqvae_params = (
            list(policy.vqbet.action_head.vqvae_model.encoder.parameters())
            + list(policy.vqbet.action_head.vqvae_model.decoder.parameters())
            + list(policy.vqbet.action_head.vqvae_model.vq_layer.parameters())
        )
        decay_params, no_decay_params = policy.vqbet.policy.configure_parameters()
        decay_params = (
            decay_params
            + list(policy.vqbet.rgb_encoder.parameters())
            + list(policy.vqbet.state_projector.parameters())
            + list(policy.vqbet.rgb_feature_projector.parameters())
            + [policy.vqbet._action_token]
            + list(policy.vqbet.action_head.map_to_cbet_preds_offset.parameters())
        )

        if cfg.policy.sequentially_select:
            decay_params = (
                decay_params
                + list(policy.vqbet.action_head.map_to_cbet_preds_primary_bin.parameters())
                + list(policy.vqbet.action_head.map_to_cbet_preds_secondary_bin.parameters())
            )
        else:
            decay_params = (
                decay_params
                + list(policy.vqbet.action_head.map_to_cbet_preds_bin.parameters())
            )

        optim_groups = [
            {
                "params": decay_params,
                "weight_decay": cfg.training.adam_weight_decay,
                "lr": cfg.training.lr,
            },
            {
                "params": vqvae_params,
                "weight_decay": 0.0001,
                "lr": cfg.training.vqvae_lr,
            },
            {
                "params": no_decay_params,
                "weight_decay": 0.0,
                "lr": cfg.training.lr,
            },
        ]
        super(VQBeTOptimizer, self).__init__(
            optim_groups,
            cfg.training.lr,
            cfg.training.adam_betas,
            cfg.training.adam_eps,
        )
```

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+
+        self.param_groups = self.encoder_optimizer.param_groups
+
+    def step(self):
```
Reflected. Please see the comment above:)

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+            self.bet_optimizer1.zero_grad()
+            self.bet_optimizer2.zero_grad()
+
+class VQBeTScheduler(nn.Module):
```
Reflected. I put all the parameters for phase 2 in the scheduler. However, to make sure that the scheduler don't changes the `lr` during phase 1, we couldn't delete this part at all. If you have better idea for this function, please let me know:)
```
class VQBeTScheduler(nn.Module):
    def __init__(self, optimizer, cfg):
        super().__init__()
        # VQ-BeT use scheduler only for rgb encoder. Since we took rgb encoder part from diffusion policy, we also follow the same scheduler from it.
        from diffusers.optimization import get_scheduler
        self.n_vqvae_training_steps = cfg.training.n_vqvae_training_steps
        self.optimizing_step = 0

        self.lr_scheduler = get_scheduler(
            cfg.training.lr_scheduler,
            optimizer=optimizer,
            num_warmup_steps=cfg.training.lr_warmup_steps,
            num_training_steps=cfg.training.offline_steps,
        )


    def step(self):
        self.optimizing_step +=1
        if self.optimizing_step >= self.n_vqvae_training_steps:
            self.lr_scheduler.step()
```

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+            self.lr_scheduler.step()
+
+class VQBeTRgbEncoder(nn.Module):
+    """Encoder an RGB image into a 1D feature vector.
```
Thank you:)
Maybe we should change the same comment in `class DiffusionRgbEncoder(nn.Module):` too.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+
+        self.train()
+
+    def toggle_discretized(self, state=True):
```
dropped this part.

jayLEE0301 replied:
First I dropped this part, but to delete `def train(self, mode=True):` part below, we have to make sure that EMA parts in RVQ is not updated when simply passing data through them. The most simple way is setting `freeze_codebook = True`, and thus, the following two lines should be executed simultaneously.

```
self.discretized = torch.tensor(True)
self.vq_layer.freeze_codebook = True
```

this parts are required (a) when we load a trained model, and (b) when the rvq train is finished. Therefore, didn't removed this part
```
def toggle_discretized(self, state=True):
    self.discretized = torch.tensor(state)
    self.vq_layer.freeze_codebook = state
```

Do you have better idea for this?

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+            hidden_channels=[config.vqvae_enc_hidden_dim, config.vqvae_enc_hidden_dim, self.config.output_shapes["action"][0] * self.config.action_chunk_size],
+        )
+
+        self.train()
```
I deleted `self.train()` in this part.

There are three remaining `self.eval()` in `modeling_vqbet.py`:
- in `def select_action(self, batch: dict[str, Tensor])`
- `def load_state_dict(self, *args, **kwargs)` in `class VqVae(nn.Module)`
- `def train(self, mode=True)` in `class VqVae(nn.Module)`

Regarding first `self.eval()`, `def select_action(self, batch: dict[str, Tensor])` in `modeling_act.py` has same thing, but `modeling_diffusion.py`  don't have `self.eval()` inside it.

After testing with trained checkpoint, I found that the first and second `self.eval()` doesn't affects the results, so deleted.

jayLEE0301 replied:
Removed all `self.train()` and `self.eval()` (some of them was deleted by removing `def load_state_dict` and `def train`

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+    def toggle_discretized(self, state=True):
+        self.discretized = torch.tensor(state)
+
+    def check_discretized(self):
```
Good idea. I removed this part:)

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+        self.encoder.eval()
+        self.decoder.eval()
+
+    def train(self, mode=True):
```
Removed, and now prevents EMA updates through `self.vq_layer.freeze_codebook = True`

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+        return rep_loss, metric
+
+
+    def load_state_dict(self, *args, **kwargs):
```
Removed `def load_state_dict`, and now `discretized` which was in `def toggle_discretized`, and `vq_layer.freeze_codebook` is now neatly included in the model state dict.

Technically, we could consolidate `discretized` and `vq_layer.freeze_codebook`, but I am not doing that; this is because vq_layer.freeze_codebook is in a very low-level subclass, and we don't want to reach that far down every time we check whether it's discretized.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+        else:
+            return einops.rearrange(output, "N (T A) -> N T A", A=self.config.output_shapes["action"][0])
+
+    def preprocess(self, state):
```
removed this function since it is only using `einops.rearrange`.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+        else:
+            self.eval()
+
+    def draw_logits_forward(self, encoding_logits):
```
Thank you for your opinion:) I removed `def draw_logits_forward` since it is not used, and changed `def draw_code_forward` to `def get_embeddings_from_code`

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2460 @@
+
+
+
+class VqVae(nn.Module):
```
I added some comments on for this class:)

jayLEE0301 commented:
Thank you for the review!

I've resolved all the comments. In high-level view,

- I removed and consolidated all the redundant `optimizers` and `schedulers` (merged all the optimizers for phase 1 and phase 2 into one, leaving only one optimizer). I also deleted `def step`, `def zero_grad`
- removed redundant functions, and now using original `load_state_dict`, `train`, and `eval` of `nn.module`. In fact, the existence of these functions was to prevent the EMA updates after RVQ training has ended. This is now implemented instead via `self.vq_layer.freeze_codebook = torch.tensor(True)` and `torch.no_grad()`
- Added comments and changed some confusing function names, and removed unused parts.

jayLEE0301 commented:
Regarding consolidating optimizers, I found that the loss of vqvae learning parts does not drop if we put all the parameters in single optimizer. Do you have any recommendations for 

- if we put all the parameters in single optimizer and scheduler (same as current PR version): purple curve
- if we separate vqvae optimizer and bet optimizer/scheduler (the code attached below): blue curve


reconstruction error for vqvae learning:
<img width="207" alt="Screenshot 2024-06-16 at 10 23 28 PM" src="https://github.com/huggingface/lerobot/assets/30570922/752e156a-842d-4598-a7cd-cc49c33e7637">

Do you have any ideas on how to integrate the optimizers without running into this issue, or any insight into why it's causing this issue?

```
class VQBeTOptimizer(nn.Module):
    def __init__(self, policy, cfg):
        super().__init__()
        self.n_vqvae_training_steps = cfg.training.n_vqvae_training_steps
        self.optimizing_step = 0


        vqvae_params = (
            list(policy.vqbet.action_head.vqvae_model.encoder.parameters())
            + list(policy.vqbet.action_head.vqvae_model.decoder.parameters())
            + list(policy.vqbet.action_head.vqvae_model.vq_layer.parameters())
        )
        self.vqvae_optimizer = torch.optim.Adam(
            vqvae_params, lr=cfg.training.vqvae_lr, weight_decay=0.0001
        )
        decay_params, no_decay_params = policy.vqbet.policy.configure_parameters()
        decay_params = (
            decay_params
            + list(policy.vqbet.rgb_encoder.parameters())
            + list(policy.vqbet.state_projector.parameters())
            + list(policy.vqbet.rgb_feature_projector.parameters())
            + [policy.vqbet._action_token]
            + [policy.vqbet._code_to_token]
            + list(policy.vqbet.action_head.map_to_cbet_preds_offset.parameters())
        )

        if cfg.policy.sequentially_select:
            decay_params = (
                decay_params
                + list(policy.vqbet.action_head.map_to_cbet_preds_primary_bin.parameters())
                + list(policy.vqbet.action_head.map_to_cbet_preds_secondary_bin.parameters())
            )
        else:
            decay_params = (
                decay_params
                + list(policy.vqbet.action_head.map_to_cbet_preds_bin.parameters())
            )

        optim_groups = [
            {
                "params": decay_params,
                "weight_decay": cfg.training.adam_weight_decay,
                "lr": cfg.training.lr,
            },
            {
                "params": no_decay_params,
                "weight_decay": 0.0,
                "lr": cfg.training.lr,
            },
        ]
        self.second_optimizer = torch.optim.Adam(
            optim_groups,
            cfg.training.lr,
            cfg.training.adam_betas,
            cfg.training.adam_eps,
        )
        self.param_groups = self.second_optimizer.param_groups
    def step(self):
        self.optimizing_step +=1
        # pretraining VQ-VAE (Training Phase 1)
        if self.optimizing_step < self.n_vqvae_training_steps:
            self.vqvae_optimizer.step()
        # training BeT (Training Phase 2)
        else:
            self.second_optimizer.step()

    def zero_grad(self):
        # pretraining VQ-VAE (Training Phase 1)
        if self.optimizing_step < self.n_vqvae_training_steps:
            self.vqvae_optimizer.zero_grad()
        # training BeT (Training Phase 2)
        else:
            self.second_optimizer.zero_grad()

class VQBeTScheduler(nn.Module):
    def __init__(self, optimizer, cfg):
        super().__init__()
        # VQ-BeT use scheduler only for rgb encoder. Since we took rgb encoder part from diffusion policy, we also follow the same scheduler from it.
        from diffusers.optimization import get_scheduler
        self.n_vqvae_training_steps = cfg.training.n_vqvae_training_steps
        self.optimizing_step = 0

        self.lr_scheduler = get_scheduler(
            cfg.training.lr_scheduler,
            optimizer=optimizer.second_optimizer,
            num_warmup_steps=cfg.training.lr_warmup_steps,
            num_training_steps=cfg.training.offline_steps,
        )


    def step(self):
        self.optimizing_step +=1
        if self.optimizing_step >= self.n_vqvae_training_steps:
            self.lr_scheduler.step()
```

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,1172 @@
+
+
+
+def pretrain_vqvae(vqvae_model, n_vqvae_training_steps, actions):
```
nit: this appears to be used once in `VQBeTHead.discretize` and is called within a method that wraps a single line. What do you say we:
1. Stay in line with the OOP approach and put this in `VQBeTHead`
2. Put the logic in `discretize` as I'm not sure the indirection adds value.

jayLEE0301 replied:
I deleted this code and put inside `VQBeTHead`.
Also, I think we don't need `discretize` function in `VQBeTModel`, so deleted it and made it use `self.vqbet.action_head.discretize` directly in `VQBeTPolicy`

alexander-soare commented:
@jayLEE0301 I found the issue with the optimizer. The single optimizer is fine on its own but because of the way you were using the scheduler, it was keeping the learning rate at 0 (due to the linear warmup phase). Specifically because of:

```python
if self.optimizing_step >= self.n_vqvae_training_steps:
    self.lr_scheduler.step()
```

So, I suppose to address this, we could:

a) Don't use warmup, but this would change the training behaviour so we'd need to test that it doesn't have a negative impact.
b) Implement the scheduler ourselves so we have a flat LR to start with, followed by warmup, then cosine:

(b) would be great as we remove dependency on diffusers. Heres' what they do

```
    def lr_lambda(current_step):
        if current_step < num_warmup_steps:
            return float(current_step) / float(max(1, num_warmup_steps))
        progress = float(current_step - num_warmup_steps) / float(max(1, num_training_steps - num_warmup_steps))
        return max(0.0, 0.5 * (1.0 + math.cos(math.pi * float(num_cycles) * 2.0 * progress)))

    return LambdaLR(optimizer, lr_lambda, last_epoch)
```

we can just add one more conditional branch to achieve what we want.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,2359 @@
+class VQBeTScheduler(nn.Module):
+    def __init__(self, optimizer, cfg):
+        super().__init__()
+        # VQ-BeT use scheduler only for rgb encoder. Since we took rgb encoder part from diffusion policy, we also follow the same scheduler from it.
```
Thank you:) I deleted it. let me put some proper comments after we fix the optimizer/scheduler issue.

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,1173 @@
+            return einops.rearrange(output, "N (T A) -> N T A", A=self.config.output_shapes["action"][0])
+
+    def get_code(self, state):
+        # in phase 2 of VQ-BeT training, we need a `GT code` to calculate the Focal loss for code prediction head.
```
What does GT stance for? I couldn't find the answer in the paper.

jayLEE0301 replied:
Thank you:) I changed the explanation:
```
# in phase 2 of VQ-BeT training, we need a `actual labels of action data` to calculate the Focal loss for code prediction head. (please refer to section 3.3 in the paper https://arxiv.org/pdf/2403.03181)
```
Would it be better to understand for the overall loss calculation?

alexander-soare replied:
Ah yes so I thought you maybe meant "ground truth". I think using the term "ground truth" is fine (not "GT" though) and probably more concise and legible than "actual labels"

jayLEE0301 replied:
Ok. I changed `actual labels` -> `ground truth`

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,1173 @@
+
+    def discretize(self, n_vqvae_training_steps, actions):
+        if self.vqvae_model.config.action_chunk_size == 1:
+            # not using action chunk
```
It already maintains a dummy dimension for action chunk (batch, action_chunk_size = 1, action_dim), even though this code is not using action chunk.

To clarify, I added some comments on this line.

alexander-soare replied:
I would think that the right way to do this is drop the conditional branching and just use the logic in the `else`. Then when the chunk size is 1, it should resolve itself. I verified this is true by trying both branches with chunk size = 1. Please can you also verify it?

jayLEE0301 replied:
Thank you:) I also checked that only using the logic in the `else` is fine. Removed this branching.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,1173 @@
+        self.num_images = len([k for k in config.input_shapes if k.startswith("observation.image")])
+        # This action query token is used as a prompt for querying action chunks. Please refer to "A_Q" in the image above.
+        # Note: During the forward pass, this token is repeated as many times as needed. The authors also experimented with initializing the necessary number of tokens independently and observed inferior results.
+        self._action_token = nn.Parameter(torch.randn(1, 1, self.config.gpt_input_dim))
```
I agree. removed `_` on this variable name.

jayLEE0301 commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,1173 @@
+            features[:, -len_additional_action_token:]
+        ], dim=1)
+        # pass through action head
+        pred_action = self.action_head(features)
```
changed the name `pred_action` -> `action_head_output`

alexander-soare commented on lerobot/common/policies/vqbet/configuration_vqbet.py:
```diff
@@ -0,0 +1,149 @@
+    Args:
+        n_obs_steps: Number of environment steps worth of observations to pass to the policy (takes the
+            current step and additional steps going back).
+        n_action_pred_token: Number of future tokens that VQ-BeT predicts.
```
If I understand correctly, this is not just future tokens but also includes the "current" token. Right?

jayLEE0301 replied:
Thank you, I changed this comment:
`n_action_pred_token: Total number of current token and future tokens that VQ-BeT predicts.`

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,1173 @@
+            actions = torch.cat(slices, dim=0)
+
+
+        actions = actions.to(get_device_from_parameters(self.vqvae_model))
```
Why a device transfer now? I though we should assume we are getting the actions on the correct device?

jayLEE0301 replied:
I checked that we don't need this line. Deleted this line.

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,1173 @@
+        self._focal_loss_fn = FocalLoss(gamma=2.0)
+
+    def discretize(self, n_vqvae_training_steps, actions):
+        if self.vqvae_model.config.action_chunk_size == 1:
```
nit: can we use `self.config` instead of `self.vqvae_model.config`? Same for the rest below in this function.

jayLEE0301 replied:
changed `self.config` instead of `self.vqvae_model.config`

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,1173 @@
+            # using action chunk
+            slices = []
+            slices.extend([actions[:, j:j+self.vqvae_model.config.action_chunk_size, :] for j in range(actions.shape[1]+1-self.vqvae_model.config.action_chunk_size)])
+            actions = torch.cat(slices, dim=0)
```
Could we simplify this?

1. We don't need to make an empty slices than extend it. with a list. We could have just created the list!
2. (nit) We could avoid polluting the namespace with `slices` and just do the list-comprehension in `torch.cat`. This is a style choice I'll leave to you. It depends on how easy it is to read once you format it.
3. As I mentioned above, we can use `self.config` instead of `self.vqvae_model.config`, right?

jayLEE0301 replied:
I removed `slices` and just created it using list. Does it seems better?

```
actions = torch.cat(
        [
            actions[:, j : j + self.config.action_chunk_size, :]
            for j in range(actions.shape[1] + 1 - self.config.action_chunk_size)
        ],
        dim=0
    )
```

alexander-soare commented on lerobot/common/policies/vqbet/modeling_vqbet.py:
```diff
@@ -0,0 +1,936 @@
+                    for j in range(actions.shape[1] + 1 - self.vqvae_model.config.action_chunk_size)
+                ]
+            )
+            # dimensions : (batch, action_chunk_size, action_dim)
```
I think this comment may be misleading. Shouldn't it be
```suggestion
            # dimensions : (batch * n_chunks, action_chunk_size, action_dim)
```

alexander-soare replied:
In fact, it might help to explain what is going on here. We have a batch with an action sequence that is longer than the chunk size. We are making use of this by grabbing all possible chunks of actions from the batch. We do this with a sliding window in the list comprehension.

jayLEE0301 replied:
Thank you! I updated the comments in this part:


```
    def discretize(self, n_vqvae_training_steps, actions):
        # Resize the action sequence data to fit the action chunk size using a sliding window approach.
        actions = torch.cat(
                [
                    actions[:, j : j + self.config.action_chunk_size, :]
                    for j in range(actions.shape[1] + 1 - self.config.action_chunk_size)
                ],
                dim=0
            )
        # `actions` is a tensor of shape (new_batch, action_chunk_size, action_dim) where new_batch is the number of possible chunks created from the original sequences using the sliding window.
```

alexander-soare commented:
@jayLEE0301 looking really good!

I ran training and it's going well so I'm satisfied. One minor thing I ran into was that the dataclass config says `n_vqvae_training_steps=3000` whereas the yaml says `n_vqvae_training_steps=20000` and I was confused about why the transition wasn't happening. To mitigate this sort of issue in general, I have this PR https://github.com/huggingface/lerobot/pull/292. Mind making sure the test on there passes (you'll need to add vqbet in the parameterization)? Btw, I can't merge that PR yet because of issues with CI.

alexander-soare reviewed (approved):
Ready to go! Thanks for all your hard work here @jayLEE0301
